### ISSUE TITLE
Add (dd, mkfs.ext3, e2fsck and resize2fs) as features of Singularity

### ISSUE THREAD

**Issue by robertlugg** [apptainer/singularity#4673]
### Version of Singularity:
```
3.3.0-1
```
### Expected behavior

I want to be able to build persistent overlays without the "system" tools necessary now.  Proposed implementation would be subcommands of `singularity overlay`.  Perhaps a `create` and `resize`:
`singularity overlay create [-y] ./my_overlay  size=500M *or* [bs=1M] [count=500]`
The `size` parameter would be mostly used.  Or the user could specify a bs and count parameter for finer control.  If there are other args to `dd`, you might consider them too.
and
`singularity overlay resize [-y]  size=700M`

of course the user could still use the OS commands described in the docs if they wished.

### Actual behavior

The current procedure is described in [the user guide](https://sylabs.io/guides/3.3/user-guide/persistent_overlays.html#usage).

### Justification

1. I am using systems that do not have those tools available (at least not to the users of Singularity).  Putting functionality inside of Singularity would enable overlays in these cases.
2. The commands to create and resize overlays, in the common case are unnecessarily complex.  By providing them in the Singularity package, they would be more accessible and more likely to be used.

**Comment by dtrudg** [comment]
To clarify on (1)... you are requesting this functionality, but `dd` or `resize2fs` etc. _would not_ be available where singularity is being run - so singularity could not call them to perform the work?

The functionality of `dd` is not a major issue, but I do not think we would want to re-implement base filesystem tools.

**Comment by robertlugg** [comment]
Correct, dd or resize2fs would not exist so Singularity could not call them.  Any implementation that gets me the functionality is fine.  So, for instance, if you could ship all the tools within your package, then that would be sufficient.  Or, ship the tools and add the commands.

Technically, I don't know if this can be done, but generally the request is to allow users to create overlays even if those tools aren't present.

I don't know how common this use case is.  If not possible/practical, I believe I can create those overlays on a system with those tools and then copy the empty overlays over.

PS: I can't use directory overlays as apparently they can't be run without root due to security concerns.

**Comment by dtrudg** [comment]
> I don't know how common this use case is. If not possible/practical, I believe I can create those overlays on a system with those tools and then copy the empty overlays over.

To my knowledge this wouldn't be at all common. I'd expect most Linux systems to have these available, and am curious what the environment you are running under is?

Bundling these types of binaries is not a trivial task given that we then become responsible for managing and releasing updates if there are security or bugfix issues. I don't think it's something we would seek to do unless there was a widespread requirement in the user community.

**Comment by robertlugg** [comment]
That's fair.  I'll close this issue.

I am using CentOS 6.6, 
Linux myhost 2.6.32-754.11.1.el6.x86_64 #1 SMP Tue Feb 26 15:38:56 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

I reconfirmed:  I found dd, but not resize2fs.  The system isn't controlled by me and I hate to introduce a dependency, so I think it's best for me to just build empty overlay files, then copy them to the deployment target.

I guess another alternative would be for you to ship overlays of various sizes (assuming that they compress to something very small).  In any case, thanks for considering this request.
